File: _includes/code/typescript/config-refs.datatypes.object.ts
import assert from 'assert';
import weaviate, { WeaviateClient } from 'weaviate-client';

// START ConfigureDataType
import { dataType } from 'weaviate-client';

// END ConfigureDataType

const client: WeaviateClient = await weaviate.connectToLocal();

await client.collections.delete('Person');

// START ConfigureDataType
// Create collection
const myCollection = await client.collections.create({
  name: 'Person',
  properties: [
    {
      name: 'name',
      dataType: dataType.TEXT,
    },
    {
      name: 'home_address',
      dataType: dataType.OBJECT,
      nestedProperties: [
        {
          name: 'street',
          dataType: dataType.OBJECT,
          nestedProperties: [
            {
              name: 'number',
              dataType: dataType.INT,
            },
            {
              name: 'name',
              dataType: dataType.TEXT,
            },
          ],
        },
        {
          name: 'city',
          dataType: dataType.TEXT,
        },
      ],
    },
    {
      name: 'office_addresses',
      dataType: dataType.OBJECT_ARRAY,
      nestedProperties: [
        {
          name: 'office_name',
          dataType: dataType.TEXT,
        },
        {
          name: 'street',
          dataType: dataType.OBJECT,
          nestedProperties: [
            {
              name: 'number',
              dataType: dataType.INT,
            },
            {
              name: 'name',
              dataType: dataType.TEXT,
            },
          ],
        },
      ],
    },
  ],
  // Other properties are omitted for brevity
});
// END ConfigureDataType

// START AddObject
const exampleObject = {
  name: 'John Smith',
  home_address: {
    street: {
      number: 123,
      name: 'Main Street',
    },
    city: 'London',
  },
  office_addresses: [
    {
      office_name: 'London HQ',
      street: { number: 456, name: 'Oxford Street' },
    },
    {
      office_name: 'Manchester Branch',
      street: { number: 789, name: 'Piccadilly Gardens' },
    },
  ],
};

const obj_uuid = await myCollection.data.insert(exampleObject);
// END AddObject

const returnedObject = await myCollection.query.fetchObjectById(obj_uuid);

assert.notEqual(obj_uuid, null);
assert.deepEqual(returnedObject?.properties, exampleObject);

client.close();

File: _includes/code/python/config-refs.datatypes.object.py
import weaviate

# START ConfigureDataType
from weaviate.classes.config import Property, DataType

# END ConfigureDataType

client = weaviate.connect_to_local()

# Delete collection if it exists
client.collections.delete(name="Person")

# START ConfigureDataType
# Create collection
my_collection = client.collections.create(
    name="Person",
    properties=[
        Property(name="name", data_type=DataType.TEXT),
        Property(
            name="home_address",
            data_type=DataType.OBJECT,
            nested_properties=[
                Property(
                    name="street",
                    data_type=DataType.OBJECT,
                    nested_properties=[
                        Property(name="number", data_type=DataType.INT),
                        Property(name="name", data_type=DataType.TEXT),
                    ],
                ),
                Property(name="city", data_type=DataType.TEXT),
            ],
        ),
        Property(
            name="office_addresses",
            data_type=DataType.OBJECT_ARRAY,
            nested_properties=[
                Property(name="office_name", data_type=DataType.TEXT),
                Property(
                    name="street",
                    data_type=DataType.OBJECT,
                    nested_properties=[
                        Property(name="name", data_type=DataType.TEXT),
                        Property(name="number", data_type=DataType.INT),
                    ],
                ),
            ],
        ),
    ],
    # Other properties are omitted for brevity
)
# END ConfigureDataType

# START AddObject
# Create an object
example_object = {
    "name": "John Smith",
    "home_address": {
        "street": {
            "number": 123,
            "name": "Main Street",
        },
        "city": "London",
    },
    "office_addresses": [
        {
            "office_name": "London HQ",
            "street": {"number": 456, "name": "Oxford Street"},
        },
        {
            "office_name": "Manchester Branch",
            "street": {"number": 789, "name": "Piccadilly Gardens"},
        },
    ],
}

obj_uuid = my_collection.data.insert(example_object)
# END AddObject

returned_obj = my_collection.query.fetch_object_by_id(obj_uuid)

assert obj_uuid is not None
assert returned_obj.properties == example_object

client.close()
